File: LeetCode_CloneGraph.java
import java.util.List;

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

public class LeetCode_CloneGraph {
  
    public Node [] visited;
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        visited = new Node[101];
        Arrays.fill(visited, null);
        Node clone = new Node(node.val);
        dfs(node, clone);

        return clone;
    }
    
    public void dfs(Node node, Node clone) {

        visited[clone.val] = clone;

        for(Node neighbor : node.neighbors) {
            if(visited[neighbor.val] == null) {
                Node n = new Node(neighbor.val);
                clone.neighbors.add(n);
                dfs(neighbor, n);
            } else {
                clone.neighbors.add(visited[neighbor.val]);
            }
        }
    }  
    
}

File: LeetCode_CloneGraph.cpp
#include <map>

using namespace std;

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    map<Node*, Node*> graph;
    Node* cloneGraph(Node* node) {
        if(node == NULL) return NULL;    

        if(graph.find(node) == graph.end()) {
            graph[node] = new Node(node -> val, {});
            
            for(auto current : node -> neighbors) {
                graph[node] -> neighbors.push_back(cloneGraph(current));
            }
        }

        return graph[node];
    }
};